### PR TITLE
Allow admins to invite users who don't have user accounts in Spotlight

### PR DESCRIPTION
**Pull request by jkeck**
Not sure if you all want to merge this in as is or let somebody build on-top of it.  At the very least I think we need to make modifications to the invite email because it's basically useless in its current form since it does not include any exhibit context.

mixed-users

inviting-user

### CODE REVIEW

File: app/controllers/spotlight/roles_controller.rb [hunk 1]
```diff
@@ -30,12 +30,36 @@ def update_all
...
+      end
+    end
+
+    def invite
+      params.require(:user)
+      params.require(:role)
+      user = ::User.invite!(email: params[:user])
+      role = Spotlight::Role.create(exhibit: current_exhibit, user: user, role: params[:role])
+      if role.save
+        redirect_to :back, notice: t(:'helpers.submit.role.updated')
```
Comment (jkeck) [inline app/controllers/spotlight/roles_controller.rb]: We may want to make this a different piece of text.

**Comment by jkeck** [comment]
Okay, I think this is ready for :eyes:.

File: config/routes.rb [hunk 2]
```diff
@@ -88,6 +88,8 @@
 
     resources :roles, path: 'users', only: [:index, :create, :destroy] do
       collection do
+        get :invite
```
Comment (cbeer) [inline config/routes.rb]: Should `invite` really use `GET`? It strikes me more as a `POST` or `PUT`.

File: config/locales/spotlight.en.yml [hunk 3]
```diff
@@ -466,10 +472,15 @@ en:
         bookmarklet: "%{application_name} widget"
       reindexing_in_progress: "Reindexing all resources"
     roles:
+      edit_fields:
+        invite_html: "This user does not yet exist. Would you lke to send them an %{link}?"
```
Comment (cbeer) [inline config/locales/spotlight.en.yml]: `s/lke/like/`?

File: config/locales/spotlight.en.yml [hunk 4]
```diff
@@ -261,6 +261,12 @@ en:
           non_repo_item: "Add non-repository item"
       new:
         header: Import items
+    invitation_mailer:
+      invitation_instructions:
+        hello: "Hello %{email}!"
+        someone_invited_you: "The Exhibits Administrator has invited to be a member of '%{exhibit_name}' at %{url}. You can accept it through the link below.</p>"
```
Comment (cbeer) [inline config/locales/spotlight.en.yml]: Since these contain HTML, should the keys be suffixed with `_html`?

File: app/controllers/spotlight/roles_controller.rb [hunk 5]
```diff
@@ -30,12 +30,39 @@ def update_all
...
+      #  app uses html status code to act on response
+      if ::User.where(email: exists_params).present?
+        render json: { message: 'User exists' }
+      else
+        render json: { message: 'User does not exist' }, status: :not_found
+      end
+    end
+
+    def invite
+      params.require(:user)
```
Comment (cbeer) [inline app/controllers/spotlight/roles_controller.rb]: Should these requires be moved into e.g. `#invite_params`, and have it return e.g. `params.permit(:user, :role)` to match the other methods?

File: app/controllers/spotlight/roles_controller.rb [hunk 6]
```diff
@@ -30,12 +30,39 @@ def update_all
       end
     end
 
+    def exists
+      # note: the messages returned are not shown to users and really only useful for debug, hence no translation necessary
+      #  app uses html status code to act on response
+      if ::User.where(email: exists_params).present?
+        render json: { message: 'User exists' }
```
Comment (cbeer) [inline app/controllers/spotlight/roles_controller.rb]: These strings are just for convenient debugging, I guess?

File: config/routes.rb (hunk 2 above)
Comment (jkeck) [inline config/routes.rb]: This interaction is triggered from an `Invite` link in a form (which is why it made sense at the time to make it a `GET` request).  Otherwise we'll need to have a button that is only triggerable via ajax to submit a `POST`/`PUT`.

I agree that it doesn't quite feel like a `GET` but the interaction is designed that way (and seems like a lot of trouble to do it otherwise, unless I'm missing something obvious).

File: app/controllers/spotlight/roles_controller.rb [hunk 7]
```diff
@@ -30,12 +30,39 @@ def update_all
       end
     end
 
+    def exists
+      # note: the messages returned are not shown to users and really only useful for debug, hence no translation necessary
+      #  app uses html status code to act on response
+      if ::User.where(email: exists_params).present?
```
Comment (cbeer) [inline app/controllers/spotlight/roles_controller.rb]: I wonder if we ought to make `::User` configurable at the engine level. Downstream applications might choose a different model name.

File: app/assets/javascripts/spotlight/check_user_existence.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      roleSelect().on('change', updateNoUserNoteLink);
+      submitButton().prop('disabled', true);
+    }
+
+    function updateNoUserNoteLink() {
+      var link = noUserNote().find('a');
+      var originalHref = link.data('inviteUrl');
+      var userName = target.val();
+      var role = roleSelect().val();
+      link.attr('href', originalHref + '?user=' + userName + '&role=' + role);
```
Comment (cbeer) [inline app/assets/javascripts/spotlight/check_user_existence.js]: Is there any sort of CGI escaping we need to do here?

File: config/locales/spotlight.en.yml (hunk 4 above)
Comment (jkeck) [inline config/locales/spotlight.en.yml]: Actually, I don't think this particular key should have HTML.  The `<p>` is in the mailer body (and this is actually causing a duplication `</p>`.  I'll remove the html and add a `_html` to the keys that do include markup.

File: app/assets/javascripts/spotlight/users.js [hunk 9]
```diff
@@ -27,13 +27,15 @@
     }
 
     function clear_errors(element) {
-      element.find('.has-error').removeClass('has-error');
-      element.find('.help-block').remove(); // Remove the error messages
+      element.find('.has-error')
```
Comment (cbeer) [inline app/assets/javascripts/spotlight/users.js]: Is this really equivalent to the old behavior? I'm not too clear on how jquery chaining works, I guess.

File: config/locales/spotlight.en.yml (hunk 3 above)
Comment (jkeck) [inline config/locales/spotlight.en.yml]: :+1:

File: app/views/devise/invitations/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<h2><%= t 'devise.invitations.edit.header' %></h2>
+
+<%= form_for resource, :as => resource_name, :url => invitation_path(resource_name), :html => { :method => :put } do |f| %>
```
Comment (cbeer) [inline app/views/devise/invitations/edit.html.erb]: Use Ruby 1.9-style hashes?

File: app/views/devise/invitations/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<h2><%= t 'devise.invitations.edit.header' %></h2>
```
Comment (cbeer) [inline app/views/devise/invitations/edit.html.erb]: What's different in this file? It looks the same as the upstream one to me?

File: app/views/devise/invitations/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,12 @@
+<h2><%= t "devise.invitations.new.header" %></h2>
```
Comment (cbeer) [inline app/views/devise/invitations/new.html.erb]: Same question here. What's customized about this?

File: app/views/spotlight/roles/_edit_fields.html.erb [hunk 13]
```diff
@@ -1,5 +1,11 @@
         <tr data-edit-for="<%= f.object.new_record? ? 'new' : f.object.id %>">
-          <td><%= f.text_field :user_key, hide_label: true %></td>
+          <td>
+            <%= f.email_field :user_key, hide_label: true, disabled: f.object.persisted?, data: { behavior: 'check-user-existence' } %>
+            <span style='display:none' class='help-block' data-behavior='no-user-note'>
```
Comment (cbeer) [inline app/views/spotlight/roles/_edit_fields.html.erb]: Is the inline `display: none` to stop a FOUC, or??

File: app/views/spotlight/roles/_edit_fields.html.erb [hunk 14]
```diff
@@ -1,5 +1,11 @@
...
+              <%= t('.invite_html', link: link_to(t('.invite_link'), spotlight.invite_exhibit_roles_path(current_exhibit), data: { :'invite-url' => spotlight.invite_exhibit_roles_path(current_exhibit) })) %>
```
Comment (cbeer) [inline app/views/spotlight/roles/_edit_fields.html.erb]: We're requiring Ruby 2.2, so I think:

```
'invite-url': spotlight.inv...
```

is valid here.

File: app/assets/javascripts/spotlight/users.js (hunk 9 above)
Comment (jkeck) [inline app/assets/javascripts/spotlight/users.js]: It is because `removeClass` is going to return the element that the class was removed on (which is the `.form-group`) and the `.help-block` is a child of that `.form-group`.

File: config/locales/spotlight.en.yml [hunk 15]
```diff
@@ -72,7 +72,7 @@ en:
       role:
         updated: "User has been updated."
         destroyed: "User has been removed."
-        batch_error: "There was a problem saving the users."
+        batch_error: "There was a problem saving the user(s)."
```
Comment (cbeer) [inline config/locales/spotlight.en.yml]: Can we use i18n's pluralization support to handle both these cases a little cleaner?

File: lib/generators/spotlight/install_generator.rb [hunk 16]
```diff
@@ -1,3 +1,4 @@
+# rubocop:disable ClassLength
```
Comment (cbeer) [inline lib/generators/spotlight/install_generator.rb]: Do you want to just disable `Metrics/ClassLength` for all generators? Or, maybe there's some sub-generators we ought to break out of here?

File: app/views/devise/invitations/new.html.erb (hunk 12 above)
Comment (jkeck) [inline app/views/devise/invitations/new.html.erb]: @peetucket care to comment?

File: app/views/devise/invitations/edit.html.erb (hunk 11 above)
Comment (jkeck) [inline app/views/devise/invitations/edit.html.erb]: @peetucket care to comment?

File: config/routes.rb (hunk 2 above)
Comment (cbeer) [inline config/routes.rb]: Isn't there magic `jquery-ujs` stuff that'll turn a simple link into a form?

File: config/locales/spotlight.en.yml (hunk 4 above)
Comment (cbeer) [inline config/locales/spotlight.en.yml]: :+1:

File: app/assets/javascripts/spotlight/users.js (hunk 9 above)
Comment (cbeer) [inline app/assets/javascripts/spotlight/users.js]: :+1:

**Comment by cbeer** [comment]
And, it looks like a rebase is in order.

File: app/controllers/spotlight/roles_controller.rb (hunk 5 above)
Comment (jkeck) [inline app/controllers/spotlight/roles_controller.rb]: :+1:

File: app/assets/javascripts/spotlight/check_user_existence.js (hunk 8 above)
Comment (jkeck) [inline app/assets/javascripts/spotlight/check_user_existence.js]: :+1:

File: app/controllers/spotlight/roles_controller.rb (hunk 7 above)
Comment (jkeck) [inline app/controllers/spotlight/roles_controller.rb]: I'm going to do this in a separate PR since it touches a lot of other code paths.

File: config/locales/spotlight.en.yml (hunk 15 above)
Comment (jkeck) [inline config/locales/spotlight.en.yml]: I'm going to punt on this and add a ticket for it for later.

File: config/routes.rb (hunk 2 above)
Comment (jkeck) [inline config/routes.rb]: :+1:

File: lib/generators/spotlight/install_generator.rb (hunk 16 above)
Comment (jkeck) [inline lib/generators/spotlight/install_generator.rb]: Perhaps I'm being dense, but how would I disable a particular rule for just generators?  I'm guessing we don't want to disable all rules for generators, but I'm unfamiliar with how we would do that with particular rules.

Also, I suppose we could break this out into separate generators.  I'm trying to think of where the logical divide between generators would be.  So do you think something like a dependencies generator (add dependencies and run their generators) and a generator that moves/copies/and injects into app-level files?

Comment (cbeer) [inline lib/generators/spotlight/install_generator.rb]: e.g.

```
Metrics/ClassLength:
  Exclude:
    - 'lib/generators/spotlight/*'
```

Yeah, I don't know what the right separation is. Maybe revisit it later?

File: config/locales/spotlight.en.yml (hunk 15 above)
Comment (cbeer) [inline config/locales/spotlight.en.yml]: :+1:

File: app/controllers/spotlight/roles_controller.rb (hunk 7 above)
Comment (cbeer) [inline app/controllers/spotlight/roles_controller.rb]: :+1:

File: lib/generators/spotlight/install_generator.rb (hunk 16 above)
Comment (jkeck) [inline lib/generators/spotlight/install_generator.rb]: Ah! I like that, I actually didn't know that syntax was possible, but makes sense!

**Comment by jkeck** [comment]
I think I have addressed everything and rebased.  Please let me know if I'm missing anything else, and thanks for all the advice!

**Comment by cbeer** [comment]
Do you want to squash any of those commits?

**Comment by jkeck** [comment]
Bah, yeah, sorry, I was struggling to figure out how to cohesively squash those down into something that made sense.  Hopefully what I did is workable.

**Comment by cbeer** [comment]
:+1:
